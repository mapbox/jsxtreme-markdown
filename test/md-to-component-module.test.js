'use strict';

const crypto = require('crypto');
const stripIndent = require('strip-indent');
const pify = require('pify');
const path = require('path');
const del = require('del');
const mkdirp = require('mkdirp');
const beautify = require('js-beautify');
const fs = require('fs');
const React = require('react');
const ReactDOMServer = require('react-dom/server');
const mdToComponentModule = require('../lib/md-to-component-module');

const prepText = text => stripIndent(text).trim();
const tmpDir = path.join(__dirname, '../tmp');

const loadOutputModule = content => {
  const filename = path.join(
    tmpDir,
    crypto.randomBytes(16).toString('hex') + '.js'
  );

  return pify(fs.writeFile)(filename, content).then(() => {
    const m = require(filename);
    return m;
  });
};

const renderComponent = (Component, props) => {
  return beautify.html(
    ReactDOMServer.renderToStaticMarkup(React.createElement(Component, props))
  );
};

describe('mdToComponentModule', () => {
  beforeAll(() => {
    return pify(mkdirp)(tmpDir);
  });

  afterAll(() => {
    return del(tmpDir);
  });

  afterEach(() => {
    return del(path.join(tmpDir, '*.*'));
  });

  test('default options', () => {
    const text = prepText(`
      ---
      title: Everything is ok
      ---

      # {{ frontMatter.title }}

      And a **special** number: {{ props.number }}.
    `);

    const code = mdToComponentModule(text);
    expect(code).toMatchSnapshot();
  });

  test('default options produce valid module', () => {
    const text = prepText(`
      ---
      title: Everything is ok
      ---

      # {{ frontMatter.title }}

      And a **special** number: {{ props.number }}.
    `);

    const code = mdToComponentModule(text);
    return loadOutputModule(code).then(Output => {
      const rendered = renderComponent(Output, {
        number: 77
      });
      expect(rendered).toMatchSnapshot();
    });
  });

  test('default options with modules in front matter', () => {
    const text = prepText(`
      ---
      title: Everything is ok
      modules:
        - "const Timer = require('./timer')"
        - "import { Watcher } from './watcher'"
      ---

      # {{ frontMatter.title }}

      Some introductory text.

      {{ <Watcher /> }}

      This paragraph includes a {{ <Timer /> }}.
    `);

    const code = mdToComponentModule(text);
    expect(code).toMatchSnapshot();
  });

  test('options.name', () => {
    const text = prepText(`
      # Title
    `);
    const options = {
      name: 'my-special-name'
    };
    const code = mdToComponentModule(text, options);
    expect(code).toMatchSnapshot();
  });

  test('options.template', () => {
    const text = prepText(`
      ---
      title: Foo
      list:
        - one
        - two
      ---

      # {{ frontMatter.title }}
    `);
    const options = {
      template: data => {
        return `${data.name}\n${JSON.stringify(data.frontMatter)}\n${data.jsx}`;
      }
    };
    const code = mdToComponentModule(text, options);
    expect(code).toMatchSnapshot();
  });

  test('mdToJsx options', () => {
    const text = prepText(`
      ---
      title: Foo
      list:
        - one
        - two
      ---

      # {# frontMatter.title #}

      {# <div className="foo"> #}
      The default {{ delimiter }} is ignored.
      {# </div> #}

      \`\`\`js
      var x = 3;
      \`\`\`
    `);
    const options = {
      delimiters: ['{#', '#}']
    };
    const code = mdToComponentModule(text, options);
    expect(code).toMatchSnapshot();
  });

  test('non-string primitives in front matter', () => {
    const text = prepText(`
      ---
      count: 7
      isHonest: false
      ---

      count: {{ frontMatter.count }}

      isHonest: {{ frontMatter.isHonest }}
    `);
    const code = mdToComponentModule(text);
    expect(code).toMatchSnapshot();
  });

  test('options.wrapper', () => {
    const text = prepText(`
      ---
      title: Everything is ok
      quantity: 834
      ---

      # {{ frontMatter.title }}

      Some introductory text. The quantity is {{ frontMatter.quantity }}

      Here is a number: {{ props.number }}
    `);
    const options = {
      wrapper: path.join(__dirname, './fixtures/wrapper.js')
    };
    const code = mdToComponentModule(text, options);
    return loadOutputModule(code).then(Output => {
      const rendered = renderComponent(Output, {
        number: 77
      });
      expect(rendered).toMatchSnapshot();
    });
  });

  test('documentation example, with wrapper front matter', () => {
    const text = prepText(`
      ---
      wrapper: '../wrapper'
      modules:
        - "const Timer = require('./timer')"
        - "import { Watcher } from './watcher'"
      title: Everything is ok
      quantity: 834
      ---

      # {{ frontMatter.title }}

      Some introductory text. The quantity is {{ frontMatter.quantity }}

      {{ <Watcher /> }}

      This paragraph includes a {{ <Timer /> }}.

      This component also accepts a "foo" prop: {{ props.foo }}
    `);
    const code = mdToComponentModule(text);
    expect(code).toMatchSnapshot();
  });
});
