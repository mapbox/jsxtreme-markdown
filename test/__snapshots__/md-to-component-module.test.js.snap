// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mdToComponentModule default options 1`] = `
"/*---
title: Everything is ok
---*/
\\"use strict\\";
const React = require(\\"react\\");

const frontMatter = {
  title: \\"Everything is ok\\"
};

class MarkdownReact extends React.PureComponent {
  render() {
    const props = this.props;
    return (
      <div>
        <h1>{frontMatter.title}</h1>
        <p>And a <strong>special</strong> number: {props.number}.</p>
      </div>
    );
  }
}

module.exports = MarkdownReact;
"
`;

exports[`mdToComponentModule default options produce valid module 1`] = `
"<div>
    <h1>Everything is ok</h1>
    <p>And a <strong>special</strong> number: 77.</p>
</div>"
`;

exports[`mdToComponentModule default options with modules in front matter 1`] = `
"/*---
title: Everything is ok
modules:
  - \\"const Timer = require('./timer')\\"
  - \\"import { Watcher } from './watcher'\\"
---*/
\\"use strict\\";
const React = require(\\"react\\");
const Timer = require(\\"./timer\\");
import { Watcher } from \\"./watcher\\";

const frontMatter = {
  title: \\"Everything is ok\\"
};

class MarkdownReact extends React.PureComponent {
  render() {
    const props = this.props;
    return (
      <div>
        <h1>{frontMatter.title}</h1>
        <p>Some introductory text.</p>
        <Watcher />
        <p>This paragraph includes a <Timer />.</p>
      </div>
    );
  }
}

module.exports = MarkdownReact;
"
`;

exports[`mdToComponentModule documentation example, with wrapper front matter 1`] = `
"/*---
wrapper: '../wrapper'
modules:
  - \\"const Timer = require('./timer')\\"
  - \\"import { Watcher } from './watcher'\\"
title: Everything is ok
quantity: 834
---*/
\\"use strict\\";
const React = require(\\"react\\");
const Timer = require(\\"./timer\\");
import { Watcher } from \\"./watcher\\";
const Wrapper = require(\\"../wrapper\\");

const frontMatter = {
  title: \\"Everything is ok\\",
  quantity: 834
};

class MarkdownReact extends React.PureComponent {
  render() {
    const props = this.props;
    return (
      <Wrapper {...props} frontMatter={frontMatter}>
        <div>
          <h1>{frontMatter.title}</h1>
          <p>Some introductory text. The quantity is {frontMatter.quantity}</p>
          <Watcher />
          <p>This paragraph includes a <Timer />.</p>
          <p>This component also accepts a \\"foo\\" prop: {props.foo}</p>
        </div>
      </Wrapper>
    );
  }
}

module.exports = MarkdownReact;
"
`;

exports[`mdToComponentModule mdToJsx options 1`] = `
"/*---
title: Foo
list:
  - one
  - two
---*/
\\"use strict\\";
const React = require(\\"react\\");

const frontMatter = {
  title: \\"Foo\\",
  list: [\\"one\\", \\"two\\"]
};

class MarkdownReact extends React.PureComponent {
  render() {
    const props = this.props;
    return (
      <div>
        <h1>{frontMatter.title}</h1>
        <div className=\\"foo\\">
          <p>The default {\\"{\\"}{\\"{\\"} delimiter {\\"}\\"}{\\"}\\"} is ignored.</p>
        </div>
        <pre><code className=\\"language-js\\">var x = 3;{\\"\\\\n\\"}</code></pre>
      </div>
    );
  }
}

module.exports = MarkdownReact;
"
`;

exports[`mdToComponentModule non-string primitives in front matter 1`] = `
"/*---
count: 7
isHonest: false
---*/
\\"use strict\\";
const React = require(\\"react\\");

const frontMatter = {
  count: 7,
  isHonest: false
};

class MarkdownReact extends React.PureComponent {
  render() {
    const props = this.props;
    return (
      <div>
        <p>count: {frontMatter.count}</p>
        <p>isHonest: {frontMatter.isHonest}</p>
      </div>
    );
  }
}

module.exports = MarkdownReact;
"
`;

exports[`mdToComponentModule options.name 1`] = `
"\\"use strict\\";
const React = require(\\"react\\");

const frontMatter = {};

class MySpecialName extends React.PureComponent {
  render() {
    const props = this.props;
    return <h1>Title</h1>;
  }
}

module.exports = MySpecialName;
"
`;

exports[`mdToComponentModule options.template 1`] = `
"MarkdownReact
{\\"title\\":\\"Foo\\",\\"list\\":[\\"one\\",\\"two\\"]}
<h1>{frontMatter.title}</h1>
"
`;

exports[`mdToComponentModule options.wrapper 1`] = `
"<div>
    <div>This is the wrapper. Here are its props:</div>
    <div>77</div>
    <div>Everything is ok</div>
    <div>834</div>
    <div>
        <h1>Everything is ok</h1>
        <p>Some introductory text. The quantity is 834</p>
        <p>Here is a number: 77</p>
    </div>
</div>"
`;
